### PR TITLE
rfc/oauth: Implement new Google OAuth flow.

### LINKED ISSUE #53522: sg rfc cannot register with Google OAuth
#### Steps to reproduce:

1. On a clean machine, run `sg rfc list`

#### Expected behavior:

The authentication flow completes, and `sg` tool has access to the RFC Google Docs folders.

#### Actual behavior:

Sign in with Google fails as offline code is now disabled.

Error:
```
You can't sign in to this app because it doesn't comply with 
Google's OAuth 2.0 policy for keeping apps secure.

You can let the app developer know that this app doesn't 
comply with one or more Google validation rules.

Learn more about this error.

If you are a developer of sg, see error details.

Error 400: invalid_request
```

#### References

- https://developers.google.com/identity/protocols/oauth2/policies#secure-response-handling

### PR DESCRIPTION
**Pull request by nelsonjr**
Implements the new Google OAuth flow following [Obtaining OAuth 2.0 access tokens](https://developers.google.com/identity/protocols/oauth2/native-app?hl=en#step-2:-send-a-request-to-googles-oauth-2.0-server).

Closes #53522.

## Test plan

Added tests for the OAuth flow:

    bazel test //dev/sg/internal/rfc:all

- All new code has a corresponding unit test
- Added unit test to existing code that was changed

### CODE REVIEW

**Comment by mucles** [comment]
i get the following error when running this locally 👀 
image

**Comment by mucles** [comment]
> ~/.sourcegraph/sg.secrets.json

Yes, I removed and reran it but still fails 
image

File: dev/sg/internal/rfc/rfc.go [hunk 1]
```diff
@@ -55,7 +61,7 @@ func getClient(ctx context.Context, config *oauth2.Config, out *std.Output) (*ht
 	if err := sec.Get("rfc", tok); err != nil {
 		// ...if it doesn't exist, open browser and ask user to give us
 		// permissions
-		tok, err = getTokenFromWeb(ctx, config, out)
+		tok, err = getTokenFromWeb(handleAuthResponse, ctx, NewTokenHandler(config), out)
```
Comment (mucles) [inline dev/sg/internal/rfc/rfc.go:64]: nit:i believe by convention ctx should be passed in as the first argument 
```suggestion
		tok, err = getTokenFromWeb( ctx,handleAuthResponse, NewTokenHandler(config), out)
```

File: dev/sg/internal/rfc/rfc.go [hunk 2]
```diff
@@ -68,23 +74,171 @@ func getClient(ctx context.Context, config *oauth2.Config, out *std.Output) (*ht
...
+
+func NewTokenHandler(config *oauth2.Config) *tokenHandlerImpl {
+	return &tokenHandlerImpl{
+		config: config,
+	}
+}
+
 // Request a token from the web, then returns the retrieved token.
-func getTokenFromWeb(ctx context.Context, config *oauth2.Config, out *std.Output) (*oauth2.Token, error) {
+func getTokenFromWeb(responseFactory authResponseHandlerFactory, ctx context.Context,
```
Comment (mucles) [inline dev/sg/internal/rfc/rfc.go:204]:
```suggestion
func getTokenFromWeb(ctx context.Context, responseFactory authResponseHandlerFactory,
```

**Review by mucles — APPROVED** [review]
just a minor nit, but this Looks good to me 👍

File: dev/sg/internal/rfc/rfc.go (hunk 1 above)
Comment (jhchabran) [reply dev/sg/internal/rfc/rfc.go:64]: @mucles that's correct

**Pull request merged**